### ISSUE TITLE
AsyncClient read appears to block

### ISSUE THREAD

**Issue by dcohen21** [korlibs/korio#41]
I have a lightweight Android app that, at its base, is essentially just spawning connections to an external server. The protocol is: android app opens tcp socket using `createTcpClient`, connects to the server, and sends a greeting; the server acknowledges. Then the connection waits (ideally without blocking a thread) for a message from the server using `AsyncClient.read(buffer, offset, len)`. These connections are spawned inside of individual coroutines on `Dispatchers.IO`, which has a max parallelism of 64 threads.

Using this same architecture and ktor aSockets, I am able to open basically as many connections as I want. Unfortunately, ktor aSockets has a major issue with switching coroutine contexts under the hood such that broken pipe exceptions (which should essentially be meaningless) are unhandleable and thus fatal on Android. (https://github.com/ktorio/ktor/issues/1237)

I found korio, and it seemed like a great alternative, so I swapped ktor out for korio. Unfortunately, I am only able to open a max number of connections equal to the number of threads allocated by `Dispatchers.IO`, which is the same behavior as when using Java blocking Sockets. Hence, it doesn't seem like Kotlin is able to suspend and swap coroutines to create more than 64 connections.

Should I be using a different method other than `AsyncClient.read()` to wait for a message from the server without blocking?

**Comment by soywiz** [comment]
Yes, that's like this. On the JVM I'm using AsynchronousSocketChannel. But on Android I'm using a plain Socket and `Dispatchers.IO`. I guess I did that to not increase the SDK version required on Android.
Did ktor solved this already?

**Comment by dcohen21** [comment]
@soywiz Ah I see. Supposedly this was a midnight addition to ktor `1.3.0`, but I have not tested it myself. I switched back to ktor and was able to work around the issue in my application by doing the following:
>This may not be applicable to every application that is running into this issue, but if you want to swallow this IOException so that the app doesn't fatally crash, you can override the thread's default exception handler with custom behavior using Thread.setDefaultUncaughtExceptionHandler() in MainActivity.onCreate().
https://github.com/ktorio/ktor/issues/1237#issuecomment-573456989
